File: frontend/src/types/game.ts
export type GameCondition =
  | "leader_cooperative"
  | "leader_competitive"
  | "leader_coopetition"
  | "no_leader_cooperative"
  | "no_leader_competitive"
  | "no_leader_coopetition";

export type InteractionMode = "leader" | "free";
export type ScoringMode = "cooperative" | "competitive" | "coopetition";

export interface GameConfig {
  condition: GameCondition;
  interactionMode: InteractionMode;
  scoringMode: ScoringMode;
  sessionId: string;
  playerId: string;
  totalTurns: number;
  turnTimeLimit: number;
}

// [수정] 백엔드 GameEngine의 zones 구조에 맞게 slots 내부 타입 변경
export interface ZoneState {
  zoneId: string;
  slots: (InstalledCard | null)[]; // 백엔드에서 [None]*5로 초기화됨
  maxSlots: number;
  nextExpected: number;
}

export interface InstalledCard {
  cardId: string;
  number: number;
  isCorrect: boolean;
  // 백엔드에서 추가로 보내주는 정보가 있다면 여기에 정의
}

export interface BoardState {
  zones: ZoneState[];
  errorCount: number;
  maxErrors: number;
}

export interface DeckState {
  totalCards: number;
  remainingCards: number;
  discardedCards: number;
  inHandCards?: number;
  installedCards?: number;
}

// [수정] PlayerState에 hand 속성 추가
export interface PlayerState {
  playerId: string;
  name: string;
  isHuman: boolean;
  hp: number;
  maxHp: number;
  individualScore: number;
  handSize: number;
  // [핵심 추가] 이 부분이 있어야 gameStore에서 에러가 나지 않습니다.
  hand: CardData[]; 
}

// [추가] 개별 카드 데이터 인터페이스
export interface CardData {
  cardId: string;
  number: number | null; // 내 카드일 경우 필터링되어 null로 올 수 있음
  zone: any;
  truth: string;
  hintHistory?: Array<{
    turn: number;
    hintType: "number" | "zone" | "truth";
    value: string;
    givenBy: string;
    message?: string;
  }>;
}

export interface GameState {
  sessionId: string;
  config: GameConfig;
  currentTurn: number;
  currentPhase: GamePhase;
  currentActorId?: string;
  board: BoardState;
  deck?: DeckState;
  players: PlayerState[];
  teamScore: number;
  teamScoreThreshold: number;
  thresholdReached: boolean;
  completionTargetScore?: number;
  completionTargetNumber?: number;
  maxTeamScore?: number;
  targetReached?: boolean;
  perfectReached?: boolean;
  teamFlow?: {
    mode: "balanced" | "cautious" | "information" | "push" | "conserve";
    label: string;
    description: string;
    source?: string;
  };
  isGameOver: boolean;
  gameOverReason?: string;
  winner?: string;
  turnStartedAt?: string;
}

export type GamePhase =
  | "waiting"
  | "agent_reporting"
  | "agent_turn"
  | "human_turn" // 에이전트 행동이 끝난 후 이 페이즈로 전환됨
  | "resolving"
  | "turn_end"
  | "game_over";

File: backend/app/schemas/game.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from app.core.conditions import GameCondition, InteractionMode, ScoringMode

class CardSchema(BaseModel):
    cardId: str
    number: Optional[int]
    zone: Optional[str]
    truth: str = "genuine"
    knownNumber: Optional[int] = None
    knownZone: Optional[str] = None
    knownTruth: Optional[str] = None

class ZoneStateSchema(BaseModel):
    zoneId: str
    slots: List[Optional[Dict[str, Any]]]
    maxSlots: int
    nextExpected: int

class BoardStateSchema(BaseModel):
    zones: List[ZoneStateSchema]
    errorCount: int
    maxErrors: int

class PlayerStateSchema(BaseModel):
    playerId: str
    name: str
    isHuman: bool
    hp: int
    maxHp: int
    individualScore: int
    handSize: int

class GameConfigSchema(BaseModel):
    condition: GameCondition
    interactionMode: InteractionMode
    scoringMode: ScoringMode
    sessionId: str
    playerId: str
    totalTurns: int
    turnTimeLimit: int

class GameStateSchema(BaseModel):
    sessionId: str
    config: GameConfigSchema
    currentTurn: int
    currentPhase: str
    board: BoardStateSchema
    players: List[PlayerStateSchema]
    teamScore: int
    teamScoreThreshold: int
    thresholdReached: bool
    completionTargetScore: int = 20
    completionTargetNumber: int = 4
    maxTeamScore: int = 20
    targetReached: bool = False
    isGameOver: bool
    gameOverReason: Optional[str] = None
